elastic/kibana
# [BE] [Conditional actions] Add query and timeframe params to RuleAction to filter alerts.

Resolves: #152357

With **Conditional Actions** we want to add some params to an action to filter the alerts it gets.
~~As we are already using the persisting alerts (aka alert-as-data) for summary actions, decided to use this filter with those actions first, the same feature will be applied to the non-summary actions with a follow on issue.~~
As this PR has already been waiting and the development for for-each type alerts is done, decided to add also that change in this PR. So, any rule that has `getSummarizedAlerts` can have an action with alertsFilter.

This PR adds the new `alertsFilter` object to RuleAction and allows the rule types that pass it to filter alerts for summary actions.

The filter has two groups of params: `query` and `timeframe`

`query` has two fields, `kql` and `dsl`, kql is to be entered by the user and `dsl` is to be generated and persisted out of `kql` in BE .
We send `dsl` to ES to filter alerts.

`timeframe` also has two fields: `days` and `hours`.
`days` is an array that represents the days of the week by their index numbers. (e.g. [1,2,3] means MON,TUE,WED) 

`hours` is a time range. e.g. `{start:'08:00', end:'17:00'}` means between 08:00 and 17:00.
As we save Datetime fileds as UTC time, we assume that this time range is in UTC format. (So we need to tell the users that the hours they enter are UTC times)

`{timeframe: { days: [1,2,3,4,5], hours: {start:'08:00', end:'17:00'}}` => Alert has to be started on : **Every weekday, at between 08:00 and 17:00**

The filter's schema is below:
```
{
  query: null | {
    kql: string;
    dsl?: string;
  };
  timeframe: null | {
    days: Array< 1 | 2 | 3 | 4 | 5 | 6 | 7>;
    hours: {
      start: string;
      end: string;
    };
    timezone: string;
  };
}
```

## To verify:

Add below code to the rule create/update call's payload in UI and create an alert generating rule with summary action.
Then expect the alerts to be filtered out from the action's output.

Modify https://github.com/elastic/kibana/blob/2a1740d035df9fc935c5ff46a7513b8d5c4998fb/x-pack/plugins/triggers_actions_ui/public/application/lib/rule_api/create.ts#L23 

To be:
```
const rewriteBodyRequest: RewriteResponseCase<RuleCreateBody> = ({
  ruleTypeId,
  actions,
  ...res
}): any => ({
  ...res,
  rule_type_id: ruleTypeId,
  actions: actions.map(({ group, id, params, frequency, alertsFilter }) => ({
    group,
    id,
    params,
    frequency: {
      notify_when: frequency!.notifyWhen,
      throttle: frequency!.throttle,
      summary: frequency!.summary,
    },
    alerts_filter: {
      query: {
        kql: 'kibana.alert.rule.name:<your rule name here>',
      },
      timeframe: {
        days: [1, 2, 4, 5],
        hours: { start: '08:00', end: '17:00' },
        timezone: 'UTC',
      },
    },
  })),
});
```

do the same for: https://github.com/elastic/kibana/blob/2a1740d035df9fc935c5ff46a7513b8d5c4998fb/x-pack/plugins/triggers_actions_ui/public/application/lib/rule_api/update.ts#L18
